File: pipeline/sponsor_register/normalise.py
"""Employer-name normalisation for sponsor-register matching."""

from __future__ import annotations

import re
import unicodedata
from typing import Iterable, List


LEGAL_SUFFIXES = {
    "ltd",
    "limited",
    "plc",
    "llp",
    "llc",
    "inc",
    "incorporated",
    "co",
    "company",
    "corp",
    "corporation",
    "cic",
    "gmbh",
    "sa",
    "bv",
}

TRAILING_CONTEXT_SUFFIXES = {
    "uk",
    "group",
    "holdings",
    "holding",
    "international",
}

PUNCTUATION_RE = re.compile(r"[^a-z0-9]+")


def _ascii_fold(value: str) -> str:
    return (
        unicodedata.normalize("NFKD", value)
        .encode("ascii", "ignore")
        .decode("ascii")
    )


def _strip_suffixes(tokens: Iterable[str]) -> List[str]:
    filtered = [token for token in tokens if token not in LEGAL_SUFFIXES]
    while len(filtered) > 1 and filtered[-1] in TRAILING_CONTEXT_SUFFIXES:
        filtered.pop()
    return filtered


def normalise_employer_name(name: str) -> str:
    """Return a conservative normalised employer name.

    The normaliser removes legal suffixes and trailing context words, but keeps
    core tokens such as a leading "UK" because those can be part of a real
    organisation name.
    """

    if not name:
        return ""

    text = _ascii_fold(name).lower()
    text = text.replace("&", " and ")
    text = text.replace("'", "")
    text = PUNCTUATION_RE.sub(" ", text)
    tokens = _strip_suffixes(text.split())
    return " ".join(tokens)

File: frontend/src/engine/normalise.ts
// Port of pipeline/sponsor_register/normalise.py.
// This file is a PORT and must stay in sync with the Python source of truth.

export const LEGAL_SUFFIXES = new Set([
  "ltd",
  "limited",
  "plc",
  "llp",
  "llc",
  "inc",
  "incorporated",
  "co",
  "company",
  "corp",
  "corporation",
  "cic",
  "gmbh",
  "sa",
  "bv",
]);

export const TRAILING_CONTEXT_SUFFIXES = new Set([
  "uk",
  "group",
  "holdings",
  "holding",
  "international",
]);

const PUNCTUATION_RE = /[^a-z0-9]+/g;

function asciiFold(value: string): string {
  // Python: unicodedata.normalize("NFKD", value).encode("ascii", "ignore")
  // NFKD decomposes accented chars into base + combining marks; encoding to
  // ascii with "ignore" then drops every non-ASCII codepoint (the combining
  // marks plus anything else outside ASCII). We replicate by NFKD-normalising
  // and then removing any codepoint > 0x7F.
  return value
    .normalize("NFKD")
    // eslint-disable-next-line no-control-regex
    .replace(/[^\x00-\x7F]/g, "");
}

function stripSuffixes(tokens: string[]): string[] {
  const filtered = tokens.filter((token) => !LEGAL_SUFFIXES.has(token));
  while (
    filtered.length > 1 &&
    TRAILING_CONTEXT_SUFFIXES.has(filtered[filtered.length - 1])
  ) {
    filtered.pop();
  }
  return filtered;
}

export function normaliseEmployerName(name: string): string {
  if (!name) {
    return "";
  }

  let text = asciiFold(name).toLowerCase();
  text = text.replace(/&/g, " and ");
  text = text.replace(/'/g, "");
  text = text.replace(PUNCTUATION_RE, " ");
  const tokens = stripSuffixes(text.split(/\s+/).filter(Boolean));
  return tokens.join(" ");
}
